designs/disruption-controls-drain-timeout.md

```diff
@@ -0,0 +1,78 @@
+# Disruption Controls - (in v1beta1+)
+
+## Motivation
+Users are requesting the ability to control how long karpenter will wait for a drain on an individual node to complete before forcefully evicting the remaining pods to decommission the node [#743](https://github.com/kubernetes-sigs/karpenter/issues/743). This supports two primary use cases.
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design piggybacks off the existing disruption-controls design, adding in a new `forceDrainAfter` field to the `v1beta/NodePool` `Disruption` spec. This field seems intuitively related to the other disruption controls that exist within the disruption block.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    forceDrainAfter: 24h || Never
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // ForceDrainAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable force draining of pods.
+    ForceDrainAfter *NillableDuration `json:"forceDrainAfter" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+The `forceDrainAfter` fields accepts a common duration string which defaults to a value of `Never`. Omitting the field results in the default value instructing the controller to wait indefinitely for pods to drain gracefully, maintaining the existing behavior.
+
+## Prior Art
+This is the field from CAPI's MachineDeployment spec which implements identical behavior.
+```
+$ kubectl explain machinedeployment.spec.template.spec.nodeDrainTimeout
+KIND:     MachineDeployment
+VERSION:  cluster.x-k8s.io/v1beta1
+
+FIELD:    nodeDrainTimeout <string>
+
+DESCRIPTION:
+     NodeDrainTimeout is the total amount of time that the controller will spend
+     on draining a node. The default value is 0, meaning that the node can be
+     drained without any time limitations. NOTE: NodeDrainTimeout is different
```

Comment:
This `NOTE` suggests that there has been some confusion with this semantic. Any learnings here @jackfrancis?